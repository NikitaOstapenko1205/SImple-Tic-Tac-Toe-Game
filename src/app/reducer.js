function checkWinner(field) {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
  ];

  for (let i = 0; i < lines.length; i++) {
    const [a, b, c] = lines[i];
    if (field[a].value && field[a].value === field[b].value && field[a].value === field[c].value) {
      return { winItem: `win${field[a].value}`, winNumbers: lines[i] };
    }
  }

  // check on a draw game
  const fieldFilteredLength = field.filter(item => item.value === '').length;
  if (!fieldFilteredLength) {
    return { winItem: 'drawGame', winNumbers: [] };
  }

  return { winItem: 'inProcess', winNumbers: [] };
}

export default (state, action) => {
  switch (action.type) {
    case 'MAKE_MOVE': {
      if (state.field[action.id].value === '') {
        const oldItem = state.currentItem;
        const newItem = oldItem === 'X' ? 'O' : 'X';
        const newField = _.set(state, `field[${action.id}].value`, oldItem);
        const { winItem, winNumbers } = checkWinner(newField.field);
        if (!winNumbers.length) {
          winNumbers.forEach(i => {
            newField.field[i].winItem = true;
            console.log(newField.field[i].winItem);
          });
        }
        return { ...newField, currentItem: newItem, gameStatus: winItem };
      }
      return { ...state };
    }
    case 'MAKE_NEW_GAME': {
      return { ...action.initialState };
    }
    default:
      throw new Error();
  }
};
