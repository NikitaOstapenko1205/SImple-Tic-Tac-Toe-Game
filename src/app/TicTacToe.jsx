import React, { useReducer } from 'react';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import Modal from '@material-ui/core/Modal';
import useStyles from './style';
import reducer from './reducer';

let classes;

const FormRow = props => {
  const { rows, makeMove } = props;
  const addId = id => () => {
    makeMove({ type: 'MAKE_MOVE', id });
  };
  return (
    <>
      {rows.map(item => (
        <Grid key={item.id} item>
          <Paper
            onClick={addId(item.id)}
            className={`row-${item.id} ${classes.paper} ${item.winItem ? classes.paperWin : ''}`}
          >
            {item.value}
          </Paper>
        </Grid>
      ))}
    </>
  );
};

const showModal = newGame => {
  return (
    <Modal
      className={classes.modal}
      aria-labelledby="simple-modal-title"
      aria-describedby="simple-modal-description"
      open
      display="flex"
    >
      <Button onClick={newGame} size="large" className={classes.button} variant="outlined">
        New Game
      </Button>
    </Modal>
  );
};

export default () => {
  const initialState = {
    field: [
      { value: '', id: 0, winItem: false },
      { value: '', id: 1, winItem: false },
      { value: '', id: 2, winItem: false },
      { value: '', id: 3, winItem: false },
      { value: '', id: 4, winItem: false },
      { value: '', id: 5, winItem: false },
      { value: '', id: 6, winItem: false },
      { value: '', id: 7, winItem: false },
      { value: '', id: 8, winItem: false }
    ],
    currentItem: 'X',
    gameStatus: 'inProcess'
  };

  const [state, dispatch] = useReducer(reducer, initialState);
  classes = useStyles();

  const newGame = () => {
    dispatch({ type: 'MAKE_NEW_GAME', initialState });
  };

  const getTitle = () => {
    if (state.gameStatus === 'inProcess') {
      return (
        <h1 className={classes.title}>
          Now player <span>{state.currentItem}</span> move
        </h1>
      );
    }
    if (state.gameStatus === 'drawGame') {
      return <h1 className={classes.title}>It`s a draw!</h1>;
    }
    return (
      <h1 className={classes.title}>
        Win player <span>{state.gameStatus.slice(-1)}</span>
      </h1>
    );
  };

  return (
    <>
      {getTitle()}
      <Grid className={classes.root} container justify="center" alignItems="center" spacing={2}>
        <Grid container justify="center" alignItems="center" item md={8} spacing={4}>
          <FormRow makeMove={dispatch} rows={state.field.filter(item => item.id < 3)} />
        </Grid>
        <Grid container justify="center" alignItems="center" item md={8} spacing={4}>
          <FormRow
            makeMove={dispatch}
            rows={state.field.filter(item => item.id > 2 && item.id < 6)}
          />
        </Grid>
        <Grid container justify="center" alignItems="center" item md={8} spacing={4}>
          <FormRow makeMove={dispatch} rows={state.field.filter(item => item.id >= 6)} />
        </Grid>
      </Grid>
      {state.gameStatus !== 'inProcess' && showModal(newGame)}
    </>
  );
};
