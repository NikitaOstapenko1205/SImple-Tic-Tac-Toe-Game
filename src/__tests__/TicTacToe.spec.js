import React from 'react';
import { mount } from 'enzyme';
import TicTacToe from '../app/TicTacToe';
import reducer from '../app/reducer';

describe('Reducer tests', () => {
  const initialState = {
    field: [
      { value: '', id: 0, winItem: false },
      { value: '', id: 1, winItem: false },
      { value: '', id: 2, winItem: false },
      { value: '', id: 3, winItem: false },
      { value: '', id: 4, winItem: false },
      { value: '', id: 5, winItem: false },
      { value: '', id: 6, winItem: false },
      { value: '', id: 7, winItem: false },
      { value: '', id: 8, winItem: false }
    ],
    currentItem: 'X',
    gameStatus: 'inProcess'
  };

  it('Check on default', () => {
    expect(reducer({}, { type: 'default' })).toEqual('default');
  });

  it('Check on start new game', () => {
    expect(reducer({}, { type: 'MAKE_NEW_GAME', initialState }).gameStatus).toBe('inProcess');
  });

  it('Check on first move, first field changed on X', () => {
    expect(reducer(initialState, { type: 'MAKE_MOVE', id: 0 }).field[0].value).toBe('X');
  });

  it('Check on second move, first field dont changed', () => {
    expect(reducer(initialState, { type: 'MAKE_MOVE', id: 0 }).field[0].value).toBe('X');
  });

  it('Check on win X', () => {
    expect(
      reducer(
        {
          field: [
            { value: 'X', id: 0, winItem: false },
            { value: 'X', id: 1, winItem: false },
            { value: '', id: 2, winItem: false },
            { value: 'O', id: 3, winItem: false },
            { value: 'O', id: 4, winItem: false },
            { value: '', id: 5, winItem: false },
            { value: '', id: 6, winItem: false },
            { value: '', id: 7, winItem: false },
            { value: '', id: 8, winItem: false }
          ],
          currentItem: 'X',
          gameStatus: 'inProcess'
        },
        { type: 'MAKE_MOVE', id: 2 }
      ).gameStatus
    ).toBe('winX');
  });

  it('Check on win O', () => {
    expect(
      reducer(
        {
          field: [
            { value: 'X', id: 0, winItem: false },
            { value: 'X', id: 1, winItem: false },
            { value: '', id: 2, winItem: false },
            { value: 'O', id: 3, winItem: false },
            { value: 'O', id: 4, winItem: false },
            { value: '', id: 5, winItem: false },
            { value: 'X', id: 6, winItem: false },
            { value: '', id: 7, winItem: false },
            { value: '', id: 8, winItem: false }
          ],
          currentItem: 'O',
          gameStatus: 'inProcess'
        },
        { type: 'MAKE_MOVE', id: 5 }
      ).gameStatus
    ).toBe('winO');
  });

  it('Check on draw', () => {
    expect(
      reducer(
        {
          field: [
            { value: 'X', id: 0, winItem: false },
            { value: 'X', id: 1, winItem: false },
            { value: 'O', id: 2, winItem: false },
            { value: 'O', id: 3, winItem: false },
            { value: 'O', id: 4, winItem: false },
            { value: 'X', id: 5, winItem: false },
            { value: 'X', id: 6, winItem: false },
            { value: 'O', id: 7, winItem: false },
            { value: '', id: 8, winItem: false }
          ],
          currentItem: 'X',
          gameStatus: 'inProcess'
        },
        { type: 'MAKE_MOVE', id: 8 }
      ).gameStatus
    ).toBe('drawGame');
  });
});

describe('Component snapshot tests', () => {
  const myComponent = mount(<TicTacToe />);

  it('Drawing initial field with right title', () => {
    expect(myComponent.find('h1').hasClass('makeStyles-title-2')).toBeTruthy();
    expect(myComponent.find('h1 span').text()).toBe('X');
  });

  it('Drawing initial field and check title snapshot', () => {
    expect(myComponent.find('h1').html()).toMatchSnapshot();
  });

  it('Make firts move', () => {
    const row0 = myComponent.find('.row-0').hostNodes();
    row0.simulate('click');
    expect(row0.text()).toBe('X');
    expect(row0.html()).toMatchSnapshot();
  });

  it('Check on win X', () => {
    for (let i = 0; i < 7; i++) {
      const row = myComponent.find(`.row-${i}`).hostNodes();
      row.simulate('click');
    }
    expect(myComponent.find('h1').text()).toBe('Win player X');
    expect(myComponent.find('h1').html()).toMatchSnapshot();
  });

  it('Check on start new game', () => {
    const newGameButton = myComponent.find('.MuiButtonBase-root').hostNodes();
    newGameButton.simulate('click');
    expect(myComponent.find('h1').text()).toBe('Now player X move');
    expect(myComponent.html()).toMatchSnapshot();
  });

  it('Check on draw', () => {
    const drawSequence = [0, 1, 2, 3, 5, 4, 6, 8, 7];
    drawSequence.forEach(item =>
      myComponent
        .find(`.row-${item}`)
        .hostNodes()
        .simulate('click')
    );
    expect(myComponent.find('h1').text()).toBe('It`s a draw!');
    expect(myComponent.find('h1').html()).toMatchSnapshot();
  });
});
