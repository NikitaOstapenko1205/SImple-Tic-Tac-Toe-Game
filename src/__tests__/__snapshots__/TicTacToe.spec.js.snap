// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshot tests Check on draw 1`] = `"<h1 class=\\"makeStyles-title-2\\">It\`s a draw!</h1>"`;

exports[`Component snapshot tests Check on start new game 1`] = `"<h1 class=\\"makeStyles-title-2\\">Now player <span>X</span> move</h1><div class=\\"MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-md-8\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-0 makeStyles-paper-5  MuiPaper-rounded\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-1 makeStyles-paper-5  MuiPaper-rounded\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-2 makeStyles-paper-5  MuiPaper-rounded\\"></div></div></div><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-md-8\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-3 makeStyles-paper-5  MuiPaper-rounded\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-4 makeStyles-paper-5  MuiPaper-rounded\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-5 makeStyles-paper-5  MuiPaper-rounded\\"></div></div></div><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-md-8\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-6 makeStyles-paper-5  MuiPaper-rounded\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-7 makeStyles-paper-5  MuiPaper-rounded\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiPaper-root MuiPaper-elevation1 row-8 makeStyles-paper-5  MuiPaper-rounded\\"></div></div></div></div>"`;

exports[`Component snapshot tests Check on win X 1`] = `"<h1 class=\\"makeStyles-title-2\\">Win player <span>X</span></h1>"`;

exports[`Component snapshot tests Drawing initial field and check title snapshot 1`] = `"<h1 class=\\"makeStyles-title-2\\">Now player <span>X</span> move</h1>"`;

exports[`Component snapshot tests Make firts move 1`] = `"<div class=\\"MuiPaper-root MuiPaper-elevation1 row-0 makeStyles-paper-5  MuiPaper-rounded\\">X</div>"`;
